File: TheWatch-iOS/Logging/WatchLogger.swift
import Foundation
import os.log
import UIKit

/// Structured logger facade — the primary API all components use.
///
/// Mirrors Serilog's API. Each log call is dispatched to the port asynchronously,
/// so callers never block on persistence.
///
/// Usage:
/// ```swift
/// WatchLogger.shared.information(
///     source: "LocationCoordinator",
///     template: "Escalated to {Mode} mode",
///     properties: ["Mode": "Emergency"],
///     correlationId: activeAlertId
/// )
/// ```
@Observable
final class WatchLogger: @unchecked Sendable {

    static let shared = WatchLogger()

    /// The active logging port — set during app configuration.
    /// Defaults to MockLoggingAdapter for safety.
    var port: (any LoggingPort)?

    /// Current user ID — set on login, cleared on logout.
    var userId: String?

    /// Stable device ID — set once at app startup.
    let deviceId: String = UIDevice.current.identifierForVendor?.uuidString ?? UUID().uuidString

    private let osLog = Logger(subsystem: "com.thewatch.app", category: "Structured")

    private init() {}

    // MARK: - Convenience Methods

    func verbose(source: String, template: String, properties: [String: String] = [:], correlationId: String? = nil) {
        log(level: .verbose, source: source, template: template, properties: properties, correlationId: correlationId)
    }

    func debug(source: String, template: String, properties: [String: String] = [:], correlationId: String? = nil) {
        log(level: .debug, source: source, template: template, properties: properties, correlationId: correlationId)
    }

    func information(source: String, template: String, properties: [String: String] = [:], correlationId: String? = nil) {
        log(level: .information, source: source, template: template, properties: properties, correlationId: correlationId)
    }

    func warning(source: String, template: String, properties: [String: String] = [:], error: Error? = nil, correlationId: String? = nil) {
        log(level: .warning, source: source, template: template, properties: properties, error: error, correlationId: correlationId)
    }

    func error(source: String, template: String, properties: [String: String] = [:], error: Error? = nil, correlationId: String? = nil) {
        log(level: .error, source: source, template: template, properties: properties, error: error, correlationId: correlationId)
    }

    func fatal(source: String, template: String, properties: [String: String] = [:], error: Error? = nil, correlationId: String? = nil) {
        log(level: .fatal, source: source, template: template, properties: properties, error: error, correlationId: correlationId)
    }

    // MARK: - Core Write

    private func log(
        level: LogLevel,
        source: String,
        template: String,
        properties: [String: String],
        error: Error? = nil,
        correlationId: String? = nil
    ) {
        var enriched = properties
        enriched["Platform"] = "iOS"
        enriched["OsVersion"] = UIDevice.current.systemVersion
        enriched["AppVersion"] = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "dev"

        let entry = LogEntry(
            level: level,
            sourceContext: source,
            messageTemplate: template,
            properties: enriched,
            exception: error.map { "\(type(of: $0)): \($0.localizedDescription)" },
            correlationId: correlationId,
            userId: userId,
            deviceId: deviceId
        )

        // Fire-and-forget to background
        Task.detached(priority: .utility) { [weak self] in
            await self?.port?.write(entry)
        }

        // Also echo to os_log for Console.app / Xcode
        let rendered = entry.renderedMessage
        switch level {
        case .verbose:     osLog.trace("[\(source)] \(rendered)")
        case .debug:       osLog.debug("[\(source)] \(rendered)")
        case .information: osLog.info("[\(source)] \(rendered)")
        case .warning:     osLog.warning("[\(source)] \(rendered)")
        case .error:       osLog.error("[\(source)] \(rendered)")
        case .fatal:       osLog.critical("[\(source)] \(rendered)")
        }
    }

    /// Flush buffered logs — call from scenePhase .background transition.
    func flush() {
        Task.detached(priority: .utility) { [weak self] in
            await self?.port?.flush()
        }
    }
}

File: TheWatch-Android/app/src/main/java/com/thewatch/app/data/logging/WatchLogger.kt
package com.thewatch.app.data.logging

import android.os.Build
import android.provider.Settings
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Structured logger facade — the primary API all components use.
 *
 * Mirrors Serilog's API: `logger.information("User {UserId} triggered SOS at {Lat},{Lng}", ...)`
 *
 * Each log call is fire-and-forget on a background coroutine,
 * so callers never block on persistence. The port handles buffering.
 *
 * Usage:
 * ```kotlin
 * @Inject lateinit var logger: WatchLogger
 *
 * logger.information(
 *     sourceContext = "LocationCoordinator",
 *     messageTemplate = "Escalated to {Mode} mode",
 *     properties = mapOf("Mode" to "Emergency"),
 *     correlationId = activeAlertId
 * )
 * ```
 */
@Singleton
class WatchLogger @Inject constructor(
    private val port: LoggingPort
) {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    /** Current user ID — set on login, cleared on logout. */
    var userId: String? = null

    /** Stable device ID — set once at app startup. */
    var deviceId: String? = null

    // ── Convenience methods per level ────────────────────────────

    fun verbose(
        sourceContext: String,
        messageTemplate: String,
        properties: Map<String, String> = emptyMap(),
        correlationId: String? = null
    ) = log(LogLevel.Verbose, sourceContext, messageTemplate, properties, null, correlationId)

    fun debug(
        sourceContext: String,
        messageTemplate: String,
        properties: Map<String, String> = emptyMap(),
        correlationId: String? = null
    ) = log(LogLevel.Debug, sourceContext, messageTemplate, properties, null, correlationId)

    fun information(
        sourceContext: String,
        messageTemplate: String,
        properties: Map<String, String> = emptyMap(),
        correlationId: String? = null
    ) = log(LogLevel.Information, sourceContext, messageTemplate, properties, null, correlationId)

    fun warning(
        sourceContext: String,
        messageTemplate: String,
        properties: Map<String, String> = emptyMap(),
        exception: Throwable? = null,
        correlationId: String? = null
    ) = log(LogLevel.Warning, sourceContext, messageTemplate, properties, exception, correlationId)

    fun error(
        sourceContext: String,
        messageTemplate: String,
        properties: Map<String, String> = emptyMap(),
        exception: Throwable? = null,
        correlationId: String? = null
    ) = log(LogLevel.Error, sourceContext, messageTemplate, properties, exception, correlationId)

    fun fatal(
        sourceContext: String,
        messageTemplate: String,
        properties: Map<String, String> = emptyMap(),
        exception: Throwable? = null,
        correlationId: String? = null
    ) = log(LogLevel.Fatal, sourceContext, messageTemplate, properties, exception, correlationId)

    // ── Core write ───────────────────────────────────────────────

    private fun log(
        level: LogLevel,
        sourceContext: String,
        messageTemplate: String,
        properties: Map<String, String>,
        exception: Throwable?,
        correlationId: String?
    ) {
        val enriched = properties.toMutableMap().apply {
            put("Platform", "Android")
            put("OsVersion", Build.VERSION.RELEASE)
            put("AppVersion", BuildConfig.VERSION_NAME ?: "dev")
        }

        val entry = LogEntry(
            level = level,
            sourceContext = sourceContext,
            messageTemplate = messageTemplate,
            properties = enriched,
            exception = exception?.let { "${it::class.simpleName}: ${it.message}\n${it.stackTraceToString().take(2000)}" },
            correlationId = correlationId,
            userId = userId,
            deviceId = deviceId
        )

        scope.launch {
            port.write(entry)
        }
    }

    /** Flush buffered logs — call from Application.onTrimMemory or background transition. */
    fun flush() {
        scope.launch { port.flush() }
    }
}
